enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+- [OpenStack](https://github.com/kubernetes/legacy-cloud-providers/blob/master/openstack/openstack_metrics.go) - [preserved](https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/cloudprovider/providers/openstack/metrics/metrics.go).
+
+`CCM` metrics will be collected by the [prometheus-operator](https://github.com/openshift/prometheus-operator) via `ServiceMonitor`. The `Service` endpoint will be secured with a generated certificate in a `Secret` by setting the `service.beta.openshift.io/serving-cert-secret-name` annotation.
+
+*Note: metrics state is a subject to change, and we will have influence upon adding back ones required by OpenShift before transition to out-of-tree*
+
+### Disaster Recovery
+
+In the original `KCM` implementation, where the `CCM` is currently hosted, is using a [semi-automated](https://docs.openshift.com/container-platform/4.5/backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.html) approach to disaster recovery, introducing the `forceRedeploymentReason` option in the managed CR, triggering the redeployment of the static pods, running the `KCM` code.
+
+`CCM` approach is simpler, as we run the `CCCMO` which manages the `Deployment`. Re-populating etcd will revert the old image, and the previous version of the `CCCMO` code will re-apply and restart the old `CCM` configuration.
+
+### Upgrade/Downgrade strategy
+
+A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.
+
+Upgrades from previous versions of OpenShift will look like:
+
+#### Post 4.7 (in-tree to out-of-tree per provider)
+
+Each provider will be released independently as and when the providers are deemed stable enough.
+
+Upgrade:
+
+- New `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` version updates the `kube-apiserver` and the `KCM` and removes the `--cloud-provider` flag. `cluster-kube-controller-manager-operator` in addition sets the `--external-cloud-volume-plugin=<some-platform>` flag on the `KCM` pods to ensure in-tree storage preservation.
+- `cluster-cloud-controller-manager-operator` starts and creates out-of-tree `CCM` resources for the provider.
+- `machine-config-operator` restart updates the `kubelet` with `--cloud-provider=external` option. At this point the `CCM` is running out-of-tree
+
+
+Downgrade:
+
+- `cluster-cloud-controller-manager-operator` de-provisions CCM resources for any platform.
+- `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` restart updates the `kube-apiserver` and the `kube-controller-manager` with the `--cloud-provider=<some-platform>` option.
```

Comment:
Is order here important? Downgrade updates operators in the same order as upgrade, i.e. api server first, then kube-controller-manager, then the rest.

I am not sure if CVO is going to remove cccmo on downgrade - it may not be present in the older release manifests.

Reply:
No, the order is not important. We merge a code located in two places into a third one. Backwards copy pasting it into two places and removing from the third. CCCMO should only deprovision it's resources. It can hang for a while.